src/main/java/com/nirima/jenkins/plugins/docker/DockerContainerWatchdog.java

```diff
@@ -0,0 +1,538 @@
+             * deleted that DockerCloud configuration while still the node was running; to clean up the mess,
+             * he manually force-removed the containers from the docker instance). 
+             * 
+             * At the end, this means that we cannot verify strictly that the containerId stored with this
+             * node, really exists on (any of the) DockerCloud instances or not. 
+             * The only option is that, as long as we see "some container having an id like this" on "some of
+             * our DockerCloud instances", then we should be careful with it, and do not delete the node (yet).
+             * Even if the containerId originates from a collision, then we would just postpone the deletion
+             * of the node up to that point in time, when that container gets deleted from the DockerContainer
+             * instance (on the next recurrence of this watchdog, we then would detect that situation and the
+             * node gets deleted automatically, as no container with the appropriate identifier could be detected
+             * anymore). 
+             * 
+             * It is to be hoped that collision on container identifiers are happening rarely enough to such that
+             * we do not run into troubles here. Otherwise, still a larger set of broken nodes could pile up...
+             */
+            
+            boolean seenOnDockerInstance = csmMerged.isContainerIdRegistered(dtn.getContainerId());
+            if (seenOnDockerInstance) {
+                // the node and the container still might have a proper mapping => ignore
+                continue;
+            }
+            
+            SlaveComputer computer = dtn.getComputer();
+            if (computer == null) {
+                // Probably the node is being closed down right now, so we shouldn't touch it.
+                continue;
+            }
+            
+            if (!computer.isOffline()) {
+                // the node is still running; we should not touch it.
+                continue;
+            }
+            
+            // the container is already gone for the node, but the node did not notice it yet properly
+            LOGGER.info("{} has container ID {}, but the container does not exist in any docker cloud. Will remove node.", dtn, dtn.getContainerId());
+
+            try {
+                removeNode(dtn);
+                executionStatistics.nodesRemoved++;
+            } catch (IOException e) {
+                LOGGER.warn(String.format("Failed to remove orphaned %s.", dtn), e);
+                executionStatistics.nodesRemovedFailed++;
+            }
+        }
+    }
+
+    private void checkForTimeout(Instant startedTimestamp) {
+        final Instant now = clock.instant();
+        Duration runtime = Duration.between(startedTimestamp, now);
+        
+        if (runtime.compareTo(PROCESSING_TIMEOUT_IN_MS) > 0) {
+            throw new WatchdogProcessingTimeout();
```

Comment:
I would guess that, in order to provide the information I requested [above](https://github.com/jenkinsci/docker-plugin/pull/658/files#diff-8933db77131fb3a413685c224649363aR162) it may be useful to store `startedTimestamp`, `now` and `PROCESSING_TIMEOUT_IN_MS` in the exception so that this information can be retrieved later for logging.
(I'm sure there's other ways of achieving the same effect; this is just one possibility)

Reply:
I think I found a simpler solution with b95006fd3e83b209e55f36630fe26658973cc27c.
However, if you don't like that, we can still go your way as suggested here.